Python/Product/PythonTools/PythonTools/Environments/EnvironmentSwitcherManager.cs

```diff
@@ -0,0 +1,238 @@
+
+namespace Microsoft.PythonTools.Environments {
+    /// <summary>
+    /// This keeps track of selection and other changes in the IDE and caches
+    /// the data to be displayed in the environment status bar switcher,
+    /// such as the list of all environments, the current environment,
+    /// and the current project.
+    /// </summary>
+    sealed class EnvironmentSwitcherManager : IVsSelectionEvents, IDisposable {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IInterpreterOptionsService _optionsService;
+        private readonly IInterpreterRegistryService _registryService;
+        private readonly IVsMonitorSelection _monitorSelection;
+        private uint _selectionEventsCookie;
+        private IVsHierarchy _previousHier;
+        private PythonProjectNode _project;
+
+        public event EventHandler EnvironmentsChanged;
+
+        public EnvironmentSwitcherManager(IServiceProvider serviceProvider) {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _optionsService = serviceProvider.GetComponentModel().GetService<IInterpreterOptionsService>();
+            _registryService = serviceProvider.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _monitorSelection = serviceProvider.GetService(typeof(IVsMonitorSelection)) as IVsMonitorSelection;
+
+            _optionsService.DefaultInterpreterChanged += OnDefaultInterpreterChanged;
+            _registryService.InterpretersChanged += OnInterpretersChanged;
+            _monitorSelection.AdviseSelectionEvents(this, out _selectionEventsCookie);
+
+            AllFactories = Enumerable.Empty<IPythonInterpreterFactory>();
+        }
+
+        /// <summary>
+        /// Returns whether a Python project, workspace or document is currently
+        /// selected / active. Used to control the visibility of the status bar
+        /// switcher.
+        /// </summary>
+        public bool IsInPythonMode { get; private set; }
+
+        /// <summary>
+        /// Interpreter factory that is currently applicable to the current
+        /// project, workspace, or loose document. This may be null even when
+        /// <see cref="IsInPythonMode"/> is <c>true</c>, for example when no
+        /// environment is found on the machine.
+        /// </summary>
+        public IPythonInterpreterFactory CurrentFactory { get; private set; }
+
+        /// <summary>
+        /// List of factories to present to the user, based on the current
+        /// project, workspace or loose document.
+        /// </summary>
+        /// <remarks>
+        /// For a project, this is the list of environments referenced by the
+        /// project. For a loose document or workspace, this is the list of all
+        /// environments found on the machine. 
+        /// </remarks>
+        public IEnumerable<IPythonInterpreterFactory> AllFactories { get; private set; }
```

Comment:
```cs
public IEnumerable<IPythonInterpreterFactory> AllFactories { get; }= Enumerable.Empty<IPythonInterpreterFactory>();
```
Up to you

Reply:
I have to keep the private setter because we update the value in `RefreshFactories`